Scope edit frame

mattmacf98 opened the pull request:
# Description
This is a prototype PR for the scoping work as part of https://viam.atlassian.net/browse/APP-9156. It enables the following features from the scope doc https://docs.google.com/document/d/1ynTWvFPwRYxAhTy9A3Fr7VcW2BmXMmmVbwYMpFZYNww/edit?tab=t.0#heading=h.tcicyojyqi6c

1. shows metadata about frame (position, orientation, geometry) and allows for editing these values via a form
<img width="245" height="449" alt="Screenshot 2025-09-23 at 3 16 29 PM" src="https://github.com/user-attachments/assets/babace58-b86c-4d66-987a-76aedf280995" />

2. allows re-parenting through drag and drop

https://github.com/user-attachments/assets/fca45ca4-3215-49da-b89d-da3508ddfa21

3. allow users to add and delete frames from components

https://github.com/user-attachments/assets/e2691b3a-23d0-4e64-b151-a9a3174710d8


# TODO
There are a couple of todos to take care of to take this from prototype to fully functional
- whenever the frames refresh, the selected 3d object gets completely replaced and this deselects it, we should enable a way to persist selected 3d objects even if they get replaced in the refresh
- right now the geo select just defaults to some hardcoded values, we should allow people to change the actual shape dimensions as well
- I did not implement proper orientation conversion or supporting the different modes we allow people to set orientation in
- I was having issues getting the useAppQuery to work to get the robotpart by id to grab the config, right now I am explicitly calling using the viam client
- all these updates change the config and then we wait for the viewer to refresh and update, we should either figure out a way to force refresh when the configs change or persist a temporary state to model what the eventual network state will be (I did this for a couple ones like position but others like parenting are trickier)
- drag and drop is quite temperamental and I have seen cases where accidental drags can cause weird things in the heirarchy (like a child being the parent of its own parent). We should robustify and maybe switch to using an established dragging library like https://github.com/isaacHagoel/svelte-dnd-action OR switch the UI to be a select dropdown to re-parent nodes.
- there is a bug currently that breaks my tree traversal when there is a frame that references a frame that does not exist or if there is a cyclical parenting loop (this one should be a quick fix)
- there are also a couple other ux tweaks around the input fields and buttons we should change
- a decent amount of refactoring (I jammed everything into useFrames for the prototype, would be worth moving some stuff out of there)
- this protoype does not go into how to update fragments using overrides yet

micheal-parks commented:
Nice!

Some thoughts:

> whenever the frames refresh, the selected 3d object gets completely replaced and this deselects it, we should enable a way to persist selected 3d objects even if they get replaced in the refresh

I think that we need an edit / monitor mode. Being able to edit while monitoring is going to cause problems. The biggest one that I'm seeing aside from what you describe is that we're also fetching live poses from the motion service. (in that `<Pose>` component). I think that if you're editing frames the robot should be in its null "T-pose" and not moving around or in the current pose that the robot is in.

> I did not implement proper orientation conversion or supporting the different modes we allow people to set orientation in
I was having issues getting the useAppQuery to work to get the robotpart by id to grab the config, right now I am explicitly calling using the viam client

This is another thing that confuses users, especially allowing people to edit quaternions. For now I'd just support orientation vector degrees.

> drag and drop is quite temperamental and I have seen cases where accidental drags can cause weird things in the heirarchy (like a child being the parent of its own parent). We should robustify and maybe switch to using an established dragging library like https://github.com/isaacHagoel/svelte-dnd-action OR switch the UI to be a select dropdown to re-parent nodes.

I'd vote for just making a select dropdown for reparenting for now. DnD is a time sink.

> there are also a couple other ux tweaks around the input fields and buttons we should change

I'd also vote for making the geometry type selector a dropdown since the list will grow.

mattmacf98 commented:
<img width="1409" height="753" alt="Screenshot 2025-10-01 at 2 30 44 PM" src="https://github.com/user-attachments/assets/add96d7f-e7e2-43f3-8638-5f866f21d945" />
<img width="1438" height="772" alt="Screenshot 2025-10-01 at 2 31 01 PM" src="https://github.com/user-attachments/assets/84eab4c7-4eb0-4ec8-b683-f4d5a1f5f0f9" />

In this latest update, I switched over to using implicit edit vs monitor mode. When a user starts to edit, they will be marked dirty and we will stop fetching updates to poses until they discard or save changes

I also updated the UI to only allow editing world objects which actually have frames (so the fake gripper geo that gets returned from GetGeometries is shown but cannot be changed)
